src/mdl_reader.rs

```diff
@@ -0,0 +1,301 @@
+};
+use uuid::Uuid;
+
+#[derive(thiserror::Error, uniffi::Error, Debug)]
+pub enum MDLReaderSessionError {
+    #[error("{value}")]
+    Generic { value: String },
+}
+
+#[derive(uniffi::Object)]
+pub struct MDLSessionManager(reader::SessionManager);
+
+impl std::fmt::Debug for MDLSessionManager {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Debug for SessionManager not implemented")
+    }
+}
+
+#[derive(uniffi::Record)]
+pub struct MDLReaderSessionData {
+    pub state: Arc<MDLSessionManager>,
+    uuid: Uuid,
+    pub request: Vec<u8>,
+    ble_ident: String,
+}
+
+#[uniffi::export]
+pub fn establish_session(
+    uri: String,
+    requested_items: HashMap<String, HashMap<String, bool>>,
+    trust_anchor_registry: Option<Vec<String>>,
+) -> Result<MDLReaderSessionData, MDLReaderSessionError> {
+    let namespaces: Result<BTreeMap<_, NonEmptyMap<_, _>>, non_empty_map::Error> = requested_items
+        .into_iter()
+        .map(|(doc_type, namespaces)| {
+            let namespaces: BTreeMap<_, _> = namespaces.into_iter().collect();
+            match namespaces.try_into() {
+                Ok(n) => Ok((doc_type, n)),
+                Err(e) => Err(e),
+            }
+        })
+        .collect();
+    let namespaces = namespaces.map_err(|e| MDLReaderSessionError::Generic {
+        value: format!("Unable to build data elements: {e:?}"),
+    })?;
+    let namespaces: device_request::Namespaces =
+        namespaces
+            .try_into()
+            .map_err(|e| MDLReaderSessionError::Generic {
+                value: format!("Unable to build namespaces: {e:?}"),
+            })?;
+
+    let registry = if let Some(r) = trust_anchor_registry {
+        let trust_anchors = r
+            .iter()
+            .map(|s| names_only_registry_from_pem(s))
+            .collect::<Result<Vec<TrustAnchor>, x509::error::Error>>()
+            .map_err(|e| MDLReaderSessionError::Generic {
+                value: format!("unable to parse trust anchors: {e:?}"),
+            })?;
+        let registry = TrustAnchorRegistry {
+            certificates: trust_anchors,
+        };
+        Some(registry)
+    } else {
+        None
+    };
+
+    let (manager, request, ble_ident) =
+        reader::SessionManager::establish_session(uri.to_string(), namespaces, registry).map_err(
+            |e| MDLReaderSessionError::Generic {
+                value: format!("unable to establish session: {e:?}"),
+            },
+        )?;
+    let manager2 = manager.clone();
+    let uuid =
+        manager2
+            .first_central_client_uuid()
+            .ok_or_else(|| MDLReaderSessionError::Generic {
+                value: "the device did not transmit a central client uuid".to_string(),
+            })?;
+
+    // TODO this must be wrong -- it's encoded as a hex string, but then advertised as raw bytes
+    let mut ble_ident = hex::encode(ble_ident);
+    ble_ident.insert_str(0, "0x");
```

Comment:
I believe that `mdl_bridge_native` was doing it this way because it is passing the entire structure as a string through JSON and decoding it on the other side. You might be able to get away with returning a Vec<u8> here and letting UniFFI deal with serialization.